src/client/QXmppOmemoManager.cpp

```diff
@@ -0,0 +1,5446 @@
+/// \param deviceId ID of the device
+///
+
+///
+/// \fn QXmppOmemoManager::deviceRemoved(const QString &jid, uint32_t deviceId)
+///
+/// Emitted when a device is removed.
+///
+/// \param jid device owners' bare JID
+/// \param deviceId ID of the device
+///
+
+///
+/// \fn QXmppOmemoManager::allDevicesRemoved()
+///
+/// Emitted when all devices are removed.
+///
+
+/// \cond
+void QXmppOmemoManager::setClient(QXmppClient *client)
+{
+    QXmppClientExtension::setClient(client);
+    client->setEncryptionExtension(this);
+
+    d->trustManager = client->findExtension<QXmppTrustManager>();
+    if (!d->trustManager) {
+        qFatal("QXmppTrustManager is not available, it must be added to the client before adding QXmppOmemoManager");
+    }
+
+    d->pubSubManager = client->findExtension<QXmppPubSubManager>();
+    if (!d->pubSubManager) {
+        qFatal("QXmppPubSubManager is not available, it must be added to the client before adding QXmppOmemoManager");
+    }
+
+    connect(d->trustManager, &QXmppTrustManager::trustLevelsChanged, this, [=](const QHash<QString, QMultiHash<QString, QByteArray>> &modifiedKeys) {
+        const auto &modifiedOmemoKeys = modifiedKeys.value(ns_omemo_2);
+        emit trustLevelsChanged(modifiedOmemoKeys);
+
+        for (auto itr = modifiedOmemoKeys.cbegin(); itr != modifiedOmemoKeys.cend(); ++itr) {
+            const auto &keyOwnerJid = itr.key();
+            const auto &keyId = itr.value();
+
+            // Emit 'deviceChanged()' only if there is a device with the key.
+            const auto &devices = d->devices.value(keyOwnerJid);
+            for (auto itr = devices.cbegin(); itr != devices.cend(); ++itr) {
+                if (itr->keyId == keyId) {
+                    emit deviceChanged(keyOwnerJid, itr.key());
+                    return;
+                }
+            }
+        }
+    });
+}
+
+bool QXmppOmemoManager::handlePubSubEvent(const QDomElement &element, const QString &pubSubService, const QString &nodeName)
+{
+    if (nodeName == ns_omemo_2_devices && QXmppPubSubEvent<QXmppOmemoDeviceListItem>::isPubSubEvent(element)) {
+        QXmppPubSubEvent<QXmppOmemoDeviceListItem> event;
+        event.parse(element);
+
+        switch (event.eventType()) {
+        // Items are published or deleted.
+        case QXmppPubSubEventBase::Items: {
+            // If there are IDs of deleted items, check for an inconsistency.
+            // Otherwise, check for published items.
+            if(const auto retractIds = event.retractIds(); !retractIds.isEmpty()) {
+                // Specific items are deleted.
+                const auto &retractedItem = event.retractIds().constFirst();
+                if (retractedItem == QXmppPubSubManager::standardItemIdToString(QXmppPubSubManager::Current)) {
+                    d->handleIrregularDeviceListChanges(pubSubService);
+                }
+            } else {
+                const auto &items = event.items();
```

Comment:
```suggestion
                const auto items = event.items();
```
items() returns by value.